core/src/main/scala/org/apache/spark/storage/BlockManager.scala

```diff
@@ -113,6 +116,47 @@ private[spark] class ByteBufferBlockData(
 
 }
 
+private[spark] class HostLocalDirManager(
+    futureExecutionContext: ExecutionContext,
+    cacheSize: Int,
+    externalBlockStoreClient: ExternalBlockStoreClient,
+    host: String,
+    externalShuffleServicePort: Int) extends Logging {
+
+  private val executorIdToLocalDirsCache =
+    CacheBuilder
+      .newBuilder()
+      .maximumSize(cacheSize)
+      .build[String, Array[String]]()
+
+  private[spark] def getCachedHostLocalDirs()
+      : scala.collection.Map[String, Array[String]] = executorIdToLocalDirsCache.synchronized {
+    import scala.collection.JavaConverters._
+    return executorIdToLocalDirsCache.asMap().asScala
+  }
+
+  private[spark] def getHostLocalDirs(
+      executorIds: Array[String],
+      callback: FutureCallback[java.util.Map[String, Array[String]]]): Unit = {
```

Comment:
Is this ever called outside of core? Because you can't use guava types across modules (because of shading).

Any reason why you didn't just go with my suggestion of using a closure that takes a `Try[java.util.Map[String, Array[String]]`? It looks much nicer in the caller (since you can use currying).

Reply:
I see. It is not used across modules but as you suggested I will change the callback type to `Try[java.util.Map[String, Array[String]] => Unit`.